pulp_rpm/app/metadata_parsing.py

```diff
@@ -0,0 +1,130 @@
+        raise TypeError("Unknown metadata compression type")
+    # TODO: zstd
+
+    with open_func(filelists_xml_path) as filelists_xml, open_func(other_xml_path) as other_xml:
+        filelists_parser = iterparse(filelists_xml, events=("start", "end"))
+        filelists_xml_iterator = iter(filelists_parser)
+
+        other_parser = iterparse(other_xml, events=("start", "end"))
+        other_xml_iterator = iter(other_parser)
+
+        # get a hold of the root element so we can clear it
+        # this prevents the entire parsed document from building up in memory
+        try:
+            filelists_root_element = next(filelists_xml_iterator)[1]
+            other_root_element = next(other_xml_iterator)[1]
+        # I know. This is a terrible misuse of SyntaxError. Don't blame the messenger.
+        except SyntaxError:
+            log.error("failed to parse XML metadata file")
+            raise
+
+        while True:
+            for event, filelists_element in filelists_xml_iterator:
+                # if we're not at a fully parsed package element, keep going
+                if event != "end":
+                    continue
+                # make this work whether the file has namespace as part of the tag or not
+                if not (
+                    filelists_element.tag == "package"
+                    or re.sub(NS_STRIP_RE, "", filelists_element.tag) == "package"
+                ):
+                    continue
+
+                break
+
+            for event, other_element in other_xml_iterator:
+                # if we're not at a fully parsed package element, keep going
+                if event != "end":
+                    continue
+                # make this work whether the file has namespace as part of the tag or not
+                if not (
+                    other_element.tag == "package"
+                    or re.sub(NS_STRIP_RE, "", other_element.tag) == "package"
+                ):
+                    continue
+
+                break
+
+            filelists_root_element.clear()  # clear all previously parsed ancestors of the root
+            other_root_element.clear()
+
+            (filelists_pkgid, files) = process_filelists_package_element(filelists_element)
+            (other_pkgid, changelogs) = process_other_package_element(other_element)
+
+            assert (
+                filelists_pkgid == other_pkgid
+            ), "Package key for filelists.xml and other.xml do not match"
+
+            yield filelists_pkgid, files, changelogs
+
+
+def process_filelists_package_element(element):
+    """Process one element from the filelists.xml file and return its parsed data.
+
+    :param element: object representing an XML block for one package's file list
+    :type  element: xml.etree.ElementTree.Element
+
+    :return:    unit key and dictionary containing keys "file" and "dir" where
+                values are full filesystem paths as strings.
+    :rtype:     tuple(dict, dict)
```

Comment:
This feels like it's a doc string from pulp2.
I'd suggest either to bring it up to date (including the docstring style) and add similar to other functions in this file, or leave descriptions only without writing about params in detail.
It's not that helpful to know the return types, the code is small and self explanatory, but maybe it's worth mentioning input parameters, e.g. that element is xml.etree.ElementTree.Element object.

Up to you, I'm again aiming for consistency :)